org.eclipse.jdt.ls.debug/src/org/eclipse/jdt/ls/debug/adapter/DebugAdapter.java

```diff
@@ -0,0 +1,606 @@
+        if (mainClass.endsWith(".java")) {
+            mainClass = mainClass.substring(0, mainClass.length() - 5);
+        }
+
+        String classpath;
+        try {
+            IJavaProject project = AdapterUtils.getJavaProject(arguments.projectName, mainClass);
+            classpath = AdapterUtils.computeClassPath(project);
+            classpath = classpath.replaceAll("\\\\", "/");
+        } catch (CoreException e) {
+            Logger.logException("Failed to resolve classpath.", e);
+            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot launch jvm.", null));
+        }
+
+        if (arguments.sourcePath == null || arguments.sourcePath.length == 0) {
+            this.sourcePath = new String[] { cwd };
+        } else {
+            this.sourcePath = new String[arguments.sourcePath.length];
+            System.arraycopy(arguments.sourcePath, 0, this.sourcePath, 0, arguments.sourcePath.length);
+        }
+
+        Logger.logInfo("Launch JVM with main class \"" + mainClass + "\", -classpath \"" + classpath + "\"");
+        
+        try {
+            this.debugSession = DebugUtility.launch(Bootstrap.virtualMachineManager(), mainClass, classpath);
+        } catch (IOException | IllegalConnectorArgumentsException | VMStartException e) {
+            Logger.logException("Launching debuggee vm exception", e);
+            return new Responses.ErrorResponseBody(new Types.Message(3001, "Cannot launch jvm.", null));
+        }
+
+        return new Responses.ResponseBody();
+    }
+
+    private Responses.ResponseBody attach(Requests.AttachArguments arguments) {
+        return new Responses.ResponseBody();
+    }
+
+    /**
+     * VS Code terminates a debug session with the disconnect request.
+     */
+    private Responses.ResponseBody disconnect() {
+        this.eventSubscriptions.forEach(subscription -> {
+            subscription.dispose();
+        });
+        
+        this.sendEvent(new Events.TerminatedEvent());
+        return new Responses.ResponseBody();
+    }
+
+    /**
+     * VS Code sends a configurationDone request to indicate the end of configuration sequence.
+     */
+    private Responses.ResponseBody configurationDone() {
+        this.debugSession.start();
+        this.eventSubscriptions.add(this.debugSession.eventHub().events().subscribe(debugEvent -> {
+            handleEvent(debugEvent);
+        }));
+        // The configuration sequence has done, then resume VM.
+        this.debugSession.resume();
+        return new Responses.ResponseBody();
+    }
+
+    private Responses.ResponseBody setFunctionBreakpoints(Requests.SetFunctionBreakpointsArguments arguments) {
+        return new Responses.ResponseBody();
+    }
+
+    private Responses.ResponseBody setBreakpoints(Requests.SetBreakpointArguments arguments) {
+        int size = arguments.breakpoints.length;
+        int[] debuggerLines = this.convertClientLineToDebugger(arguments.lines);
```

Comment:
If we have BreakpoinManager to manage the breakpoint states, then we need to delegate all the processing logic and let BreakpointManager handle the booking and calculating delta.
Logic inside adapter should be minimum.